test/e2e/cloud-provider-azure.go

```diff
@@ -0,0 +1,143 @@
+//go:build e2e
+// +build e2e
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	helmAction "helm.sh/helm/v3/pkg/action"
+	helmLoader "helm.sh/helm/v3/pkg/chart/loader"
+	helmCli "helm.sh/helm/v3/pkg/cli"
+	helmVals "helm.sh/helm/v3/pkg/cli/values"
+	helmGetter "helm.sh/helm/v3/pkg/getter"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InstallCloudProviderAzureHelmChart installs the official cloud-provider-azure helm chart
+func InstallCloudProviderAzureHelmChart(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) {
+	By("Waiting for workload cluster kubeconfig secret")
+	Eventually(func() error {
+		client := input.ClusterProxy.GetClient()
+		secret := &corev1.Secret{}
+		key := crclient.ObjectKey{
+			Name:      fmt.Sprintf("%s-kubeconfig", input.ConfigCluster.ClusterName),
+			Namespace: input.ConfigCluster.Namespace,
+		}
+		return client.Get(ctx, key, secret)
+	}, 20*time.Minute, 5*time.Second).Should(Succeed())
+	clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
+	By("Waiting for nodes to come online")
+	Eventually(func() error {
+		clientSet := clusterProxy.GetClientSet()
+		n, err := clientSet.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		if len(n.Items) == (int(*input.ConfigCluster.WorkerMachineCount) + int(*input.ConfigCluster.ControlPlaneMachineCount)) {
```

Comment:
This test doesn't run for windows currently, but if it did we would need to include `WINDOWS_WORKER_MACHINE_COUNT` here.

Reply:
Actually this brings up a good question. On this test, for example, this `WorkerMachineCount` value seems to apply to each node pool (not total number of expected worker nodes):

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/v1.2.1/test/e2e/azure_test.go#L216

And indeed, this value seems to be used to populate the named env var that kustomize uses to configure a single value which by convention is shared across all `MachinePool`s/`MachineDeployment`s:

https://github.com/kubernetes-sigs/cluster-api/blob/v1.1.3/cmd/clusterctl/client/config.go#L436

This check is really here to know that we're ready to reliably perform a `helm install` against the cluster. We can simply check for 1 Running node to do that.

This will work for a Windows-enabled cluster.

Do we want to just convert the reference "external-cloud-provider" template to have a Windows `MachineDeployment` as well? @CecileRobertMichon for thoughts as well

Reply:
IMO we should not change the existing external cloud provider template in this PR but instead change _all_ the reference templates to use external cloud provider as we had started doing in #1889 in a follow PR. That will include various windows flavors.

Reply:
Makes sense